src/main/scala/mesosphere/marathon/core/heartbeat/HeartbeatActor.scala

```diff
@@ -0,0 +1,97 @@
+package mesosphere.marathon.core.heartbeat
+
+import akka.actor._
+import scala.concurrent.duration._
+
+// HeartbeatActor monitors the heartbeat of some process and executes handlers for various conditions.
+// If an expected heartbeat is missed then execute an `onSkip` handler.
+// If X number of subsequent heartbeats are missed then execute an `onFailure` handler and become inactive.
+// Upon creation the actor is in an inactive state and must be sent a MessageActivate message to activate.
+// Once activated the actor will monitor for MessagePulse messages (these are the heartbeats).
+// The actor may be forcefully deactivated by sending it an MessageDeactivate message.
+class HeartbeatActor(config: Heartbeat.Config) extends LoggingFSM[HeartbeatInternal.State, HeartbeatInternal.Data] {
+  import Heartbeat._
+  import HeartbeatInternal._
+
+  startWith(StateInactive, DataNone)
+
+  when(StateInactive) {
+    case Event(MessageActivate(reactor, token), DataNone) =>
+      log.debug("heartbeat activated")
+      goto(StateActive) using DataActive(reactor, token)
+    case _ =>
+      stay // swallow all other event types
+  }
+
+  when(StateActive, stateTimeout = config.heartbeatTimeout) {
+    case Event(MessagePulse, data: DataActive) =>
+      stay using data.copy(missed = 0)
+
+    case Event(StateTimeout, data: DataActive) =>
+      if (data.missed + 1 >= config.missedHeartbeatsThreshold) {
+        data.reactor.onFailure
```

Comment:
So I put some logging in the heartbeat monitor. I'm happy to put it here
instead. If it lands both places then there will be somewhat redundant logs
... or is that what you're aiming for?

Even better, I could create a LoggingReactor that decorates what the caller
provides and generates appropriate logging before delegating. Thoughts?


On Fri, Jul 15, 2016 at 2:24 PM, Matthias Eichstedt <
notifications@github.com> wrote:

> In src/main/scala/mesosphere/marathon/core/heartbeat/HeartbeatActor.scala
> <https://github.com/mesosphere/marathon/pull/4069#discussion_r71018888>:
>
> > +
> > +  when(StateInactive) {
> > +    case Event(MessageActivate(reactor, token), DataNone) =>
> > +      log.debug("heartbeat activated")
> > +      goto(StateActive) using DataActive(reactor, token)
> > +    case _ =>
> > +      stay // swallow all other event types
> > +  }
> > +
> > +  when(StateActive, stateTimeout = config.heartbeatTimeout) {
> > +    case Event(MessagePulse, data: DataActive) =>
> > +      stay using data.copy(missed = 0)
> > +
> > +    case Event(StateTimeout, data: DataActive) =>
> > +      if (data.missed + 1 >= config.missedHeartbeatsThreshold) {
> > +        data.reactor.onFailure
>
> I'd put a log.warning here.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mesosphere/marathon/pull/4069/files/718c11edd708bc282deea7cca641f663168b5c7c#r71018888>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACPVLCuZ852NxN6Fd_CYgGC3d56C9IB5ks5qV9BAgaJpZM4JGoyy>
> .
>